sklearn/linear_model/huber.py

```diff
@@ -0,0 +1,93 @@
+from math import exp
+
+import numpy as np
+
+from scipy import optimize
+
+from sklearn.base import BaseEstimator, RegressorMixin
+from sklearn.linear_model.base import center_data, LinearModel
+from sklearn.preprocessing import StandardScaler
+from sklearn.utils import check_X_y, check_array
+
+def _huber_loss_and_gradient(w, X, y, epsilon, alpha):
+    """
+    Calculate the robust huber loss as described in
+    "A robust hybrid of lasso and ridge regression.
+
+    """
+    sigma = w[-1]
+    w = w[:-1]
+
+    # Calculate the values where |y - X'w / exp(sigma)| > epsilon
+    # The values above this threshold are outliers.
+    linear_loss = y - np.dot(X, w)
+    abs_linear_loss = np.abs(linear_loss)
+    outliers_true = abs_linear_loss * exp(-sigma) > epsilon
+
+    # Calculate the linear loss due to the outliers.
+    # This is equal to (2 * M * |y - X'w / exp(sigma)| - M**2)*exp(sigma)
+    n_outliers = np.count_nonzero(outliers_true)
+    outliers = abs_linear_loss[outliers_true]
+    outlier_loss = 2 * epsilon * np.sum(outliers) - exp(sigma) * n_outliers * epsilon**2
+
+    # Calculate the quadratic loss due to the non-outliers.-
+    # This is equal to |(y - X'w)**2 / exp(2*sigma)|*exp(sigma)
+    non_outliers = linear_loss[~outliers_true]
+    squared_loss = exp(-sigma) * np.dot(non_outliers, non_outliers)
+
+    # Calulate the gradient
+    n_samples, n_features = X.shape
+    grad = np.zeros(n_features + 1)
+
+    # Gradient due to the squared loss.
+    grad[:n_features] = 2 * exp(-sigma) * np.dot(non_outliers, -X[~outliers_true, :])
+
+    # Gradient due to the linear loss.
+    outliers_true_pos = np.logical_and(linear_loss >= 0, outliers_true)
+    outliers_true_neg = np.logical_and(linear_loss < 0, outliers_true)
+    grad[:n_features] -= 2 * epsilon * X[outliers_true_pos, :].sum(axis=0)
+    grad[:n_features] += 2 * epsilon * X[outliers_true_neg, :].sum(axis=0)
+
+    # Gradient due to the penalty.
+    grad[:n_features] += alpha * 2 * w
+
+    # Gradient due to sigma.
+    grad[-1] = n_samples * exp(sigma)
```

Comment:
Since we add 0.5 in the square loss, we also have to add 0.5 here to make the loss effectively the same as paper.

```python
grad[-1] = 0.5 * n_samples * exp(sigma)
grad[-1] -= 0.5 * n_outliers * epsilon**2 * exp(sigma)
grad[-1] -= squared_loss

return 0.5 * X.shape[0] * exp(sigma) + squared_loss + outlier_loss + alpha * np.dot(w, w), grad
```